### PR TITLE
Build py3-pylint package

### PR DESCRIPTION
**Pull request by amaltaro**
This PR builds python3 pylint specs, all the other specs come together as dependencies.

### CODE REVIEW

**Comment by muhammadimranfarooqi** [comment]
-1 to f7aba206ada9ed702558b14488bc2957e1d5da0d on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6386/test_f7aba20-20201109-104504.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Log provided is incomplete. Just in case, here is the full log:
```
vocms055:/build/dmwmbld/srv/state/dmwmbld/webarea/6386/test_f7aba20-20201109-104504.log 
```

**Comment by muhammadimranfarooqi** [comment]
It takes some time for logs to load completely in the url. Please check after few minutes

**Comment by amaltaro** [comment]
You're right, it's complete to me now.

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, from my searching in google, I found two possible reasons for that:
1) python not built with openssl
2) a `libssl-dev` package (.deb) that needs to be installed where we do pip install.

I guess 1) cannot be, because otherwise you would not be able to build python3 packages for CMSSW. Would you know if any additional seed needs to be installed on the node where we run `cmsBuild`?

**Comment by smuzaffar** [comment]
@amaltaro , you need to update openssl version. If you try to build python3 locally then you will see the following message in log
```
Could not build the ssl module!
Python requires an OpenSSL 1.0.2 or 1.1 compatible libssl with X509_VERIFY_PARAM_set1_host().
LibreSSL 2.6.4 and earlier do not provide the necessary APIs, https://github.com/libressl-portable/portable/issues/381

```

**Comment by amaltaro** [comment]
And this upgrade suddenly becomes much harder :-(
I guess the options left are either a) upgrade openssl (and affect all the other COMP services); or b) downgrade python back to 3.6.2... I guess I will try the openssl first.
Thanks again!

**Comment by smuzaffar** [comment]
have you considered dropping openssl and using it from system? For new archs of cmssw , we have dropped it and use it from system

**Comment by amaltaro** [comment]
That's an option too, but I'm afraid I don't know all the possible implications and it would affect the whole COMP stack as well. I saw that Cent OS8 is supposed to be released in production in April 2021, maybe that will be a good moment to drop openssl(?)

I already managed to build CMSSW openssl (1.0.2d) and 200 other specs. Now debugging very old specs that failed to be built with an error like:
```
* The action "build-external+cherrypy+5.4.0-comp3" was not completed successfully because Failed to build cherrypy. Log file in /build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/cherrypy/5.4.0-comp3/log. Final lines of the log file:

* You can choose a different installation directory, i.e., one that is
on PYTHONPATH or supports .pth files

* You can add the installation directory to the PYTHONPATH environment
variable.  (It must then also be on PYTHONPATH whenever you run
Python and want to use the package(s) you are installing.)

* You can set up the installation directory to support ".pth" files by
using one of the approaches described here:

https://setuptools.readthedocs.io/en/latest/easy_install.html#custom-installation-locations


Please make the appropriate changes for your system and try again.
error: Bad exit status from /build/amaltaro/w/tmp/rpm-tmp.E0O2VV (%install)


RPM build errors:
Bad exit status from /build/amaltaro/w/tmp/rpm-tmp.E0O2VV (%install)
```

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, I'm trying to build py3-pylint with a new openssl (different branch) and python3 packages fail to be built. Reason seems to be pretty clear:
```
Starting to process package external+py3-pylint+1.9.4
Download command: cd /build/amaltaro/w/tmp/tmp442Rj2; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp3/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pylint==1.9.4; /bin/ls | grep -v files.list > files.list
Checking if py3-pylint is cached.
Package external+py3-pylint+1.9.4 not found in repository. Queuing for build.
Error while downloading sources from pip://pylint/1.9.4?pip_options=--no-deps%20--no-binary%3D:all:&pip=pip3&output=/source.tar.gz.
```

it tries to use libraries from python2 pip instead of python3. It isn't clear to me where that environment comes from and I'm using the HEAD of pkgtools V00-32-xx. Could you point me into the right direction?

**Comment by smuzaffar** [comment]
For python3 , you should use pylint version 2.6.0. Always check in cmssw branch (e.g. https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_2_X/master/pip/requirements.txt ) what we are doing for few of these py3 packages :-)

**Comment by amaltaro** [comment]
@smuzaffar Thanks Shahzad, I'll get those versions updated.
However, isn't the download command wrong:
```
Download command: cd /build/amaltaro/w/tmp/tmp442Rj2; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp3/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pylint==1.9.4; /bin/ls | grep -v files.list > files.list
```
?

I mean, it sources py2-pip environment instead of py3-pip(?)

**Comment by smuzaffar** [comment]
that is correct, let me see

**Comment by smuzaffar** [comment]
are you sure that you are using latest pkgtools? If I try to build then I see this error (which uses the correct py3-pip)
```
Download command: cd /build/muz/dd/w/tmp/tmpgn4CPe; . /build/muz/dd/w/slc7_amd64_gcc630/external/py3-pip/19.3.1/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pylint==1.9.4; /bin/ls | grep -v files.list > files.list
Fetching sources for py3-pylint
Error while downloading sources from pip://pylint/1.9.4?pip_options=--no-deps%20--no-binary%3D:all:&pip=pip3&package_dependency=py3-pip&output=/source.tar.gz.

```

**Comment by amaltaro** [comment]
Yep, here is what I have in my area:
```
amaltaro@vocms055:/build/amaltaro/pkgtools $ git log -5 --pretty=oneline --decorate
9ac50cd782f34eaa8c688a128fb481d15ac0f5fd (HEAD, V00-32-XX) addlow upload dependnecies via UPLOAD_DEPENDENCIES flag in spec
2c01f518091dbfef8af5ebef511e3982b35b9a32 only use hash tag is tag is actually needed
```

There must be a problem with `build-with-pip3.file` then. When we adopted it from CMSSW, we decided to break it into scripts (one for py2 and the other for py3). I couldn't spot the problem yet though.

File: build-with-pip3.file [hunk 1]
```diff
@@ -31,7 +39,7 @@ Requires: python3 py3-setuptools
 BuildRequires: py3-pip
 
 %if "%{?source0:set}" != "set"
-%define source0 pip://%{pip_name}/%{realversion}?pip_options=%{PipDownloadOptions}&pip=pip3&output=/%{pkgsource}
+%define source0 pip://%{pip_name}/%{realversion}?pip_options=%{PipDownloadOptions}&pip=pip3&package_dependency=py3-pip&output=/%{pkgsource}
```
Comment (smuzaffar) [inline build-with-pip3.file]: this change here looks good and this should instruct cmsBuild to use py3-pip env for downloading.

**Comment by smuzaffar** [comment]
build-with-pip3 looks good. I tried building py3-pylint using your branch and it correctly used py-pip3. The other issue is that you need to update openssl otherwise pip3 fails due to SSL/TLS issue

File: build-with-pip3.file (hunk 1 above)
Comment (amaltaro) [reply build-with-pip3.file]: Sigh... sorry for the noise. I did not have this change in the branch I'm currently testing! You're the master Shahzad! Thanks

**Comment by amaltaro** [comment]
It builds properly to me too. Thanks Shahzad.

**Comment by muhammadimranfarooqi** [comment]
-1 to 346261bde5bea612a44d0a724b7bfbe33461bd58 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6386/test_346261b-20201111-152615.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
This is expected to fail because we need to first bump openssl to a newer version (I have things coming up today).

**Comment by muhammadimranfarooqi** [comment]
+1 to ae553dae54284a4e458f3c1d01e5bcd4639c008f on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6386/test_ae553da-20210127-151004.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
@muhammadimranfarooqi  given that this PR touches only the python3 specs, can you please merge it? Thanks

**Pull request merged**